ggml-opencl.c

```diff
@@ -188,48 +186,143 @@ static cl_program build_program_from_source(cl_context ctx, cl_device_id dev, co
-    clGetDeviceIDs(platform, CL_DEVICE_TYPE_ALL, num_devices, devices, NULL);
-    device = devices[dev_num];
-    char device_buffer[1024];
-    clGetDeviceInfo(device, CL_DEVICE_NAME, sizeof(device_buffer), &device_buffer, NULL);
-    printf("Using Platform: %s Device: %s\n", platform_buffer, device_buffer);
-    context = clCreateContext(NULL, 1, &device, NULL, NULL, &err);
-    CL_CHECK(err, "clCreateContext");
-    queue = clCreateCommandQueue(context, device, CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE, &err);
-    CL_CHECK(err, "clCreateCommandQueue");
-
-    free(platforms);
-    free(devices);
-
-    program = build_program_from_source(context, device, clblast_dequant);
+    CL_CHECK(clGetPlatformIDs(NPLAT, platforms, &num_platforms));
+
+    char * GGML_OPENCL_PLATFORM = getenv("GGML_OPENCL_PLATFORM");
+    if (GGML_OPENCL_PLATFORM != NULL) {
+        unsigned plat_num;
+        if (sscanf(GGML_OPENCL_PLATFORM, " %u", &plat_num) == 1) {
+            if (plat_num >= num_platforms) {
+                fprintf(stderr, "ggml_opencl: There is no platform %d\n", plat_num);
+                exit(1);
+            } else {
+                platform = platforms[plat_num];
+                CL_CHECK(clGetPlatformInfo(platform, CL_PLATFORM_NAME, sizeof(text_buffer), &text_buffer, NULL));
+            }
+        } else {
+            for (unsigned i = 0; i < num_platforms; i++) {
+                CL_CHECK(clGetPlatformInfo(platforms[i], CL_PLATFORM_NAME, sizeof(text_buffer), &text_buffer, NULL));
+                if (strstr(text_buffer, GGML_OPENCL_PLATFORM) != NULL) {
+                    platform = platforms[i];
+                    break;
+                }
+            }
+        }
+        if (platform == NULL) {
+            fprintf(stderr, "ggml_opencl: no platform matching '%s' was found.\n", GGML_OPENCL_PLATFORM);
+            exit(1);
+        } else {
+            fprintf(stderr, "ggml_opencl: selecting platform: '%s'\n", text_buffer);
+        }
+    }
+
+    char * GGML_OPENCL_DEVICE = getenv("GGML_OPENCL_DEVICE");
+    if (GGML_OPENCL_DEVICE != NULL) {
+        cl_device_id devices[NDEV];
+        cl_uint num_devices;
+        CL_CHECK(clGetDeviceIDs(platform, CL_DEVICE_TYPE_ALL, NDEV, devices, &num_devices));
+
+        unsigned dev_num;
+        if (sscanf(GGML_OPENCL_DEVICE, " %u", &dev_num) == 1) {
+            if (dev_num >= num_devices) {
+                fprintf(stderr, "ggml_opencl: There is no device %d\n", dev_num);
+                exit(1);
+            } else {
+                device = devices[dev_num];
+                CL_CHECK(clGetDeviceInfo(device, CL_DEVICE_NAME, sizeof(text_buffer), &text_buffer, NULL));
+            }
+        } else {
+            for (unsigned i = 0; i < num_devices; i++) {
+                CL_CHECK(clGetDeviceInfo(devices[i], CL_DEVICE_NAME, sizeof(text_buffer), &text_buffer, NULL));
+                if (strstr(text_buffer, GGML_OPENCL_DEVICE) != NULL) {
```

Comment:
If I don't select platform or GPU, it works:
`ggml_opencl: using device: 'NVIDIA GeForce RTX 3060'`

It doesn't find `3060` for some reason, not sure why. `RTX` works. Doesn't find `40` either, but `P40` works. Some limitation with numbers?
```
ggml_opencl: selecting platform: 'NVIDIA CUDA'
ggml_opencl: There is no device 3060
```

Reply:
Hmm, yes, it converts it to a number but there isn't a 3061st device available.

Need to rethink this.

Reply:
Oh, that makes sense. Keeping numbered platforms/devices as an option isn't a bad idea, I just didn't think of it.

Reply:
Maybe just make it clear in the error that it didn't find device number 3060, that would be enough for me.

Reply:
I'm rewriting the whole thing to be more clearer and avoid duplication.